web_app/contract_tools/ekubo_research.py

```diff
@@ -0,0 +1,238 @@
+"""
+Ekubo Flash Loan Research and Implementation Guide
+================================================
+
+This research document outlines how to implement flash loans on Ekubo AMM platform,
+based on their documentation and implementation examples.
+
+Key Concepts:
+------------
+1. Flash Accounting: Ekubo's internal token balance accounting system.
+2. Free Flash Loans: No fees for borrowing within same transaction.
+3. Atomic Execution: All operations must complete in single transaction.
+
+Contract Reference:
+------------------
+https://docs.ekubo.org/integration-guides/reference/contract-addresses#governance-contracts
+
+Documentation: https://docs.ekubo.org/
+
+Implementation Flow:
+------------------
+API Endpoints(https://mainnet-api.ekubo.org/openapi.json) and Data Structures
+"""
+
+from dataclasses import dataclass
+from typing import List, Dict, Optional, Tuple
+from decimal import Decimal
+import httpx
+from pydantic import BaseModel
+
+@dataclass
+class PoolKey:
+    """Ekubo pool identification structure"""
+    token0: str
+    token1: str
+    fee: int
+    tick_spacing: int
+    extension: str
+
+@dataclass
+class SwapRoute:
+    """Structure for defining swap path"""
+    pool_key: PoolKey #Dict[str, str] - Contains token0, token1, fee, tick_spacing, extension
+    sqrt_ratio_limit: int
+    skip_ahead: int
+
+@dataclass
+class TokenAmount:
+    """Token amount with direction"""
+    token: str
+    amount: int
+    is_positive: bool
+
+class RouteQuote(BaseModel):
+    """Single route quote from API"""
+    amount: str
+    priceImpact: str
+    route: List[Dict]
+
+class QuoteResponse(BaseModel):
```

Comment:
There was a validation error while running your code. Have you tried running it? You can add a screenshot of the output and test setup(in the same file, will be enough) to verify everything works well.